### PR TITLE
[SYCL][Doc] Add overview of kernel-program caching

### PR DESCRIPTION
**Pull request by s-kanaev**

### CODE REVIEW

File: sycl/doc/KernelProgramCache.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# A brief overview of kernel/program caching mechanism.
+
+The cache is employed when one submits kernel for execution or builds program or
```
Comment (kbobrovs) [inline sycl/doc/KernelProgramCache.md:3]: I find it quite hard to follow the narration, frankly. Can we somewhat restruct the text?
Intro
  - what caches SYCL runtime implements (what is cached)
  - why (performance? why low performance w/o caches)
  - cache key and cache value (conceptually, w/o synchronization details at this point)
  - when caching happens, when cached value is used
  - mention as TODO that there is no eviction (https://github.com/intel/llvm/issues/2517)
  - why synchronization is needed
Details
  - details on when caching is skipped, when it is done.
  - overall synchronization scheme - what resource is guarded, what algorithm and primitives are used,
    what is sync primitive-to-resource correspondence (1-1, 1-many).
 - would be good to gave a picture with threads, cache elements and sync primitives

**Review by kbobrovs — CHANGES_REQUESTED** [review]

File: sycl/doc/KernelProgramCache.md (hunk 1 above)
Comment (s-kanaev) [reply sycl/doc/KernelProgramCache.md:111]: I believe the readability of the document is improved now.

**Review by alexbatashev — CHANGES_REQUESTED** [review]
This patch is going to break our automation. @s-kanaev please update https://github.com/intel/llvm/blob/sycl/sycl/doc/index.rst accordingly. Also I'm not sure if HTML tags are going to work with Sphinx. Please, try building documentation locally. You can figure out how it's done in CI from https://github.com/intel/llvm/blob/sycl/.github/workflows/gh_pages.yml

**Comment by s-kanaev** [comment]
@alexbatashev I've built documentation with `docs-sycl-html` target. The file looks good and the links are OK in generated html. The new file is added to table of contents.

**Review by alexbatashev — APPROVED** [review]
index.rst LGTM

File: sycl/doc/KernelProgramCache.md [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+The cache stores underlying PI objects of `cl::sycl::program` and
+`cl::sycl::kernel` in a per-context data storage. The storage consists of two
+maps: one is for programs and the other is for kernels.
+
+Programs mapping's key consists of three components:
+kernel set id<sup>[1](#what-is-ksid)</sup>, specialized constants, device this
+program is built for.
+
+Kernels mapping's key consists of three components too: program the kernel
```
Comment (kbobrovs) [inline sycl/doc/KernelProgramCache.md:80]:
```suggestion
The kernel map's key consists of three components too: program the kernel
```

File: sycl/doc/KernelProgramCache.md [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+belongs to, kernel name<sup>[2](#what-is-kname)</sup>, device the program is
+built for.
+
+<a name="what-is-ksid">1</a>: Kernel set id is merely a number of translation
+unit which contains at least one kernel.
```
Comment (kbobrovs) [inline sycl/doc/KernelProgramCache.md:85]: I think this is inaccurate. It is a rather a ordinal number of the device binary image the kernel is contained in. Kernels from multiple TUs may end-up in the same binary image.

File: sycl/doc/KernelProgramCache.md [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+<a name="what-is-kname">2</a>: Kernel name is mangled class name which is
```
Comment (kbobrovs) [inline sycl/doc/KernelProgramCache.md:87]:
```suggestion
<a name="what-is-kname">2</a>: Kernel name is a mangled kernel ID class' name which is
```

File: sycl/doc/KernelProgramCache.md [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+ - Employ the same built object for multiple devices of the same ISA,
+   capabilities and so on. *NOTE:* It's not really known if it's possible to
+   check if two distinct devices are *exactly* the same.
+ - Improve testing: cover real use-cases. See currently covered cases [here](https://github.com/intel/llvm/blob/sycl/sycl/unittests/kernel-and-program/Cache.cpp).
+
+
+## Implementation details
+
+The caches are represented with instance of [`KernelProgramCache`](https://github.com/intel/llvm/blob/sycl/sycl/source/detail/kernel_program_cache.hpp)
+class. The class is instantiated in a per-context manner.
```
Comment (kbobrovs) [inline sycl/doc/KernelProgramCache.md:105]:
```suggestion
class. The runtime creates on instance of the class per distinct SYCL context. (A context object which is a result of copying another context object is not "distinct", as it corresponds to the same underlying internal object representing a context.)
```

File: sycl/doc/KernelProgramCache.md [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+The `KernelProgramCache` is the storage descrived above.
```
Comment (kbobrovs) [inline sycl/doc/KernelProgramCache.md:107]:
```suggestion
The `KernelProgramCache` is essentially a pair of maps as described above.
```

File: sycl/doc/KernelProgramCache.md [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+
+### When does the cache come at work?
+
+The cache is employed when one submits kernel for execution or builds program or
```
Comment (kbobrovs) [inline sycl/doc/KernelProgramCache.md:112]:
```suggestion
The cache is used when one submits a kernel for execution or builds a program or
```

File: sycl/doc/KernelProgramCache.md [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+kernel with SYCL API. That means that the cache works when either user
```
Comment (kbobrovs) [inline sycl/doc/KernelProgramCache.md:113]:
```suggestion
with SYCL API. That means that the cache works when either user
```

File: sycl/doc/KernelProgramCache.md [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+explicitly calls `program::build_with_kernel_type<>()`/`program::get_kernel<>()`
+methods or SYCL RT builds or gets the required kernel. Cacheability of an object
```
Comment (kbobrovs) [inline sycl/doc/KernelProgramCache.md:115]:
```suggestion
methods or SYCL RT builds a program or gets the required kernel as needed during application execution. Cacheability of an object
```

File: sycl/doc/KernelProgramCache.md [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+is verified with `program_impl::is_cacheable()` method. SYCL RT will check if
```
Comment (kbobrovs) [inline sycl/doc/KernelProgramCache.md:116]:
```suggestion
can be tested with `program_impl::is_cacheable()` method. SYCL RT will check if
```

File: sycl/doc/KernelProgramCache.md [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+program is cacheable and will get the kernel with call to
+`ProgramManager::getOrCreateKernel()` method.
```
Comment (kbobrovs) [inline sycl/doc/KernelProgramCache.md:118]:
```suggestion
only try to insert cacheable programs or kernels into the cache. This is done as a part of `ProgramManager::getOrCreateKernel()` method.
```

File: sycl/doc/KernelProgramCache.md [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+
+*NOTE:* a kernel is only cacheable if and only if the program it belongs to is
+cacheable. On the other hand if the program is cacheable, then each and every
+kernel of this program will be cached also.
+
+
+Invoked by user `program::build_with_kernel_type<>()` and
+`program::get_kernel<>()` methods will call either
+`ProgramManager::getBuildPIProgram()` or `ProgramManager::getOrCreateKernel()`
+method respectively. Now, both these methods will call template
```
Comment (kbobrovs) [inline sycl/doc/KernelProgramCache.md:129]:
```suggestion
All requests to build a program or to create a kernel - whether they originate from explicit user API calls or from internal SYCL runtime execution logic - end up with calling the
```

File: sycl/doc/KernelProgramCache.md [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+function [`getOrBuild()`](https://github.com/intel/llvm/blob/sycl/sycl/source/detail/program_manager/program_manager.cpp#L149)
+with multiple lambdas passed to it:
```
Comment (kbobrovs) [inline sycl/doc/KernelProgramCache.md:131]:
```suggestion
with a number of lambda functions passed as arguments:
```

File: sycl/doc/KernelProgramCache.md [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+function. The rest of operation is done with the help of atomics and condition
+variables (plus a mutex for proper work of condition variable).
+
+*GetCache* function returns a reference to mapping `key->value` out of locked
+instance of cache. We will see rationale behind it a bit later.
+
+*Build* function actually builds the kernel or program.
+
+Caching isn't done:
+ - when program is built out of source i.e. with
```
Comment (kbobrovs) [inline sycl/doc/KernelProgramCache.md:148]:
```suggestion
 - when program is built out of source with
```

File: sycl/doc/KernelProgramCache.md [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+   `program::build_with_source()` or `program::compile_with_source()` method;
+ - when program is result of linking of multiple programs.
```
Comment (kbobrovs) [inline sycl/doc/KernelProgramCache.md:150]:
```suggestion
 - when the program is a result of linking multiple programs.
```

**Review by kbobrovs — CHANGES_REQUESTED** [review]
First portion, reviewed until Thread safety.

File: sycl/doc/KernelProgramCache.md [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+
+### Core of caching mechanism
+
+Now, how `getOrBuild` works?
+First, we fetch the cache with sequential calls to Acquire and GetCache
+functions. Then, we check if we're the first ones who build this kernel/program.
+This is achieved with attempt to insert another key-value pair into the map.
+At this point we try to insert `BuildResult` stub instance with status equal to
+"in progress" which will allow other threads to know that someone is (i.e.
+we're) building the object (i.e. kernel or program) now. If insertion fails we
```
Comment (rakeshkrishnaiyer) [inline sycl/doc/KernelProgramCache.md:190]: If insertion fails, we

File: sycl/doc/KernelProgramCache.md [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+will wait for building thread to finish with call to `waitUntilBuilt` function.
+This function will throw stored exception<sup>[2](#exception-data)</sup> upon
+build failure. This allows waiting threads to result the same as the building
```
Comment (rakeshkrishnaiyer) [inline sycl/doc/KernelProgramCache.md:193]: This allows waiting threads to see the same result as the building thread.

File: sycl/doc/KernelProgramCache.md [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+thread. Special case of the failure is when build result doesn't contain the
+error (i.e. the error wasn't of `cl::sycl::exception` type) and the pointer to
+object in `BuildResult` instance is nil. In this case the building thread has
+finished build process and returned the user an error. Though, this error could
+be of spurious/sporadic nature. Hence, the waiting thread will try to build the
```
Comment (rakeshkrishnaiyer) [inline sycl/doc/KernelProgramCache.md:198]: In this case, the building thread has finished the build process and has returned an error to the user. But this error may be sporadic in nature and may be spurious. Hence, the waiting thread will

File: sycl/doc/KernelProgramCache.md [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+same object once more.
+
+`BuildResult` structure also contains synchronization objects: mutex and
+condition variable. We employ them to signal waiting threads that the build
+process for this kernl/program is finished (either successfuly or with a
+failure).
+
+
+<a name="remove-pointer">1</a>: The use of `std::remove_pointer` was omitted in
```
Comment (rakeshkrishnaiyer) [inline sycl/doc/KernelProgramCache.md:207]: was omitted for the sake of simplicity here.

File: sycl/doc/KernelProgramCache.md (hunk 17 above)
Comment (s-kanaev) [reply sycl/doc/KernelProgramCache.md:193]: Fixed.

File: sycl/doc/KernelProgramCache.md (hunk 18 above)
Comment (s-kanaev) [reply sycl/doc/KernelProgramCache.md:198]: Fixed.

File: sycl/doc/KernelProgramCache.md [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+which is provided to methods of `cl::sycl::handler` (e.g. `parallel_for` or
+`single_task`).
+
+
+## Points of improvement (things to do)
+
+ - Implement LRU policy on cached objects. See [issue](https://github.com/intel/llvm/issues/2517).
+ - Allow for caching of objects built with some build options.
+ - Employ the same built object for multiple devices of the same ISA,
+   capabilities and so on. *NOTE:* It's not really known if it's possible to
```
Comment (romanovvlad) [inline sycl/doc/KernelProgramCache.md:104]: I believe, this cannot be done in the SYCL RT because underlying APIs do not support such a thing.
It should be improved inside these implementations of these APIs instead.

Comment (s-kanaev) [reply sycl/doc/KernelProgramCache.md:104]: Noted.

File: sycl/doc/KernelProgramCache.md [hunk 21]
```diff
@@ -0,0 +1,223 @@
...
+
+<a name="what-is-program">1</a>: Here "program" means an internal SYCL runtime
+object corresponding to a SPIRV module or native binary defining a set of SYCL
+kernels and/or device functions.
+
+
+## Data structure of cache
+
+The cache stores underlying PI objects behind `cl::sycl::program` and
+`cl::sycl::kernel` user-levelobjects in a per-context data storage. The storage
```
Comment (romanovvlad) [inline sycl/doc/KernelProgramCache.md:81]:
```suggestion
`cl::sycl::kernel` user-level objects in a per-context data storage. The storage
```

**Review by romanovvlad — COMMENTED** [review]
LGTM

File: sycl/doc/KernelProgramCache.md [hunk 22]
```diff
@@ -0,0 +1,223 @@
...
+perform the rest of build/wait process without forcing other threads to wait on
+lock availability.
+
+A specialization of helper class [Locked](https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/detail/locked.hpp)
+for reference of proper mapping is returned by Acquire function. The use of this
+class implements RAII to make code look cleaner a bit. Now, GetCache function
+will return the mapping to be employed that includes the 3 components: kernel
+name, device as well as any specialization constants. These get added to
+`BuildResult` and are cached. The `BuildResult` structure is specialized with
+either `PiKernel` or `PiProgram`<sup>[1](#remove-program)</sup>.
```
Comment (pvchupin) [inline sycl/doc/KernelProgramCache.md:190]:
```suggestion
either `PiKernel` or `PiProgram`<sup>[1](#remove-pointer)</sup>.
```

File: sycl/doc/KernelProgramCache.md (hunk 21 above)
Comment (s-kanaev) [reply sycl/doc/KernelProgramCache.md:81]: Fixed.

File: sycl/doc/KernelProgramCache.md (hunk 22 above)
Comment (s-kanaev) [reply sycl/doc/KernelProgramCache.md:190]: Fixed.

**Review by romanovvlad — DISMISSED** [review]

**Comment by pvchupin** [comment]
@kbobrovs please review

File: sycl/doc/KernelProgramCache.md [hunk 23]
```diff
@@ -0,0 +1,223 @@
...
+Hence, what is cached is a wrapper structure `BuildResult` which contains three
+information fields - pointer to built resource, build error (if applicable) and
+current build status (either of "in progress", "succeeded", "failed").
+
+One can find definition of `BuildResult` template in [KernelProgramCache](https://github.com/intel/llvm/blob/sycl/sycl/source/detail/kernel_program_cache.hpp).
+
+Pointer to built resource and build result are both atomic variables. Atomicity
+of these variables allows one to hold lock on cache for quite a short time and
+perform the rest of build/wait process without forcing other threads to wait on
+lock availability.
```
Comment (kbobrovs) [inline sycl/doc/KernelProgramCache.md:182]: I think it is not the atomicity of fields which allows that, but the synchronization approach 
used. In this strategy fields in the BuildResult are accessed from multiple threads, so they must be atomic to ensure proper memory fences and exclude visibility of partially written data.

```suggestion
The built resource access synchronization approach aims at minimizing the time any thread 
holds the global lock guarding the maps to improve performance. To achieve that, the global lock is acquired only for the duration of the global map access. Actual build of the program happens outside of the lock, so other threads can request or build other programs in the meantime. A thread requesting a `BuildResult` instance via `getOrBuild` can go one of three ways:
A) Build result is **not** available, it is the first thread to request it. Current thread will then execute the build letting others wait for the result using the per-build result condition variable kept in `BuildResult`'s `MBuildCV` field.
B) Build result is **not** available, another thread is already building the result. Current thread will then wait for the result using the `MBuildCV` condition variable.
C) Build result **is** available. The thread simply takes it from the `Ptr` field w/o using any mutexes or condition variables.
As noted before, access to `BuildResult` instance fields may occur from different threads simultaneously, but the global lock is no longer held. So, to make it safe and to make sure only one thread builds the requested program, the following is done:
- program build state is reflected in the `State` field, threads use compare-and-swap technique to compete who will do the build and become thread A. Threads C will find 'DONE' in this field and immediately return the with built result at hand.
- thread A and thread(s) B use the `MBuildCV` conditional variable field and `MBuildResultMutex`  mutex field guarding that variable to implement the "single producer-multiple consumers scheme". 
- the build result itself appears in the 'Ptr' field when available.
All fields are atomic because they can accessed from multiple threads.
```

**Review by kbobrovs — CHANGES_REQUESTED** [review]

File: sycl/doc/KernelProgramCache.md (hunk 23 above)
Comment (s-kanaev) [reply sycl/doc/KernelProgramCache.md:182]: Good point. Changed.

**Review by pvchupin — APPROVED** [review]
